chapter/methodology.md

```diff
@@ -0,0 +1,1116 @@
+
+Variable usages
+  ~ can be in three different states. 
+  
+    1. `Usage`s that can not (yet) be mapped to a declaration are tagged `Unbound`
+    2. A `Resolved` usage introduces a back-link to the complementary `Declaration`
+    3. For record destructuring resolution of the name might need to be `Deferred` to the post-processing as discussed in [@sec:variable-usage-and-static-record-access].
+
+Other nodes
+  ~ of the AST that do not participate in the usage graph, are linearized as `Structure` -- A wildcard variant with no associated data.
+
+<!-- TODO: Add graphics -->
+
+#### Scopes
+
+<!-- TODO: how to explain scopes -->
+
+The Nickel language implements lexical scopes with name shadowing.
+
+1. A name can only be referred to after it has been defined
+2. A name can be redefined locally
+
+An AST inherently supports this logic.
+A variable reference always refers to the closest parent node defining the name and scopes are naturally separated using branching.
+Each branch of a node represents a sub-scope of its parent, i.e., new declarations made in one branch are not visible in the other.
+
+When eliminating the tree structure, scopes have to be maintained in order to provide auto-completion of identifiers and list symbol names based on their scope as context.
+Since the bare linear data structure cannot be used to deduce a scope, related metadata has to be tracked separately.
+The language server maintains a register for identifiers defined in every scope.
+This register allows NLS to resolve possible completion targets as detailed in [@sec:resolving-by-scope].
+
+The globally tracked scope metadata maps `ScopeId`s to a list of identifiers defined in the scope.
+An instance of the linearizer is valid for a single scope and hence corresponds to a unique `ScopeId`.
+Every item generated by the same linearizer is associated with the `ScopeId` of the instance.
+A scope branch during the traversal of the AST is indicated through the `Linearizer::scope()` method.
+The `Linearizer::scope()` method creates a new linearizer instance with a new `ScopeId`.
+A `ScopeId` in turn is a "scope path", a list of path elements where the prefix is equal to the parent scope's `ScopeId`.
+[Listing @lst:nickel-scope-example] shows the scopes for a simple expression in Nickel explictly.
+
+
+
+```{.nickel #lst:nickel-scope-example caption="Explicit display of Nickel scopes"}
+---------------------------------------------+/1
+                                             |
+let record                                   |
+  = { -------------------------------+ /1/1  |
+                                     |       |
+    key1 = "value", -------- /1/1/1  |       |
+    key2 = 123, ------------ /1/1/2  |       |
+                                     |       |
+  }----------------------------------+       |
+  in record ------------------------/1/2     |
+                                             |
+---------------------------------------------+
+```
```

Comment:
```suggestion
```{.nickel #lst:nickel-scope-example caption="Explicit display of Nickel scopes"}
---------------------------------------------+ /1
                                             |
let record                                   |
  = { -------------------------------+ /1/1  |
                                     |       |
    key1 = "value", -------- /1/1/1  |       |
    key2 = 123, ------------ /1/1/2  |       |
                                     |       |
  } ---------------------------------+       |
  in record -------------------------- /1/2  |
                                             |
---------------------------------------------+
```
```